Util/channel/src/main/java/io/deephaven/util/channel/BaseSeekableChannelContext.java

```diff
@@ -8,29 +8,36 @@
 import org.jetbrains.annotations.Nullable;
 
 import javax.annotation.OverridingMethodsMustInvokeSuper;
+import java.util.HashMap;
+import java.util.Map;
 import java.util.function.Supplier;
 
 public class BaseSeekableChannelContext implements SeekableChannelContext {
     /**
-     * An opaque resource object hosted by this context.
+     * A cache of opaque resource objects hosted by this context.
      */
-    private SafeCloseable resource;
+    private final Map<String, SafeCloseable> resourceCache = new HashMap<>();
 
     @Override
     @Nullable
-    public final SafeCloseable apply(@NotNull final Supplier<SafeCloseable> resourceFactory) {
+    public final SafeCloseable apply(final String key, @NotNull final Supplier<SafeCloseable> resourceFactory) {
+        SafeCloseable resource = resourceCache.get(key);
         if (resource == null) {
             resource = resourceFactory.get();
+            if (resource != null) {
+                resourceCache.put(key, resource);
+            }
         }
         return resource;
     }
 
     @Override
     @OverridingMethodsMustInvokeSuper
     public void close() {
-        if (resource != null) {
-            resource.close();
-            resource = null;
+        for (final SafeCloseable resource : resourceCache.values()) {
+            if (resource != null) {
```

Comment:
I'm not sure the `null` check is relevant anymore? We may want to clear the `resourceCache` (or remove as we are it.remove() while going through).

Reply:
Right now, we don't have a use case for clearing or updating the cache. 
That is why we use a `BytesArrayHolder` or a `DecompressorHolder`, because then the caller can update the holder instance to clear it or update it.